Add advanced options on Recover Wallet tab

guggero opened the pull request:
This PR adds two new "advanced options" to the tab "Recover Wallet":
* "**Account Key Path**": Useful if an existing BIP39 wallet should be imported that was created with a different software and used a different derivation path.
* "**Number of keys to generate**": Determines how many keys will be generated/derived when the wallet is created. Useful if a wallet should be imported/restored that has been used a lot already.

## Preview
![Screenshot from 2019-04-30 22-54-01](https://user-images.githubusercontent.com/1008879/56992773-16e94500-6b9b-11e9-8619-45cd8287bb0d.png)

I plan to add additional PRs with more advanced options in the "Recover Wallet" tab to support different derivation methods used by other wallet software. I hope that's in the general interest of the project since people might want to import their existing wallet rather than create a new one and send funds to it.

nopara73 commented:
First of all, I really like how you did the UX. I can really appreciate that someone doesn't add unnecessary complexity. In fact we could even steal this "Show advanced options" UX pattern and could use more heavily. Refer to the issue: https://github.com/zkSNACKs/WalletWasabi/issues/1369  

@lontivero can you review the code?

nopara73 commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletView.xaml:
```diff
@@ -29,6 +29,12 @@
+          <StackPanel Spacing="8" IsVisible="{Binding ShowAdvancedOptions}">
+            <controls:ExtendedTextBox Text="{Binding AccountKeyPath}" Watermark="Account Key Path" UseFloatingWatermark="True"/>
+            <controls:ExtendedTextBox Text="{Binding MinGapLimit}" Watermark="Number of keys to derive" UseFloatingWatermark="True"/>
+            <TextBlock Text="Warning: Deriving more than 50 keys can take a while since their UTXOs will also be synchronized."/>
```
Warning is not necessary. In principle warnings are not really needed in "Advanced" mode. But also a normal Wasabi wallet uses thousands of keys, because coinjoins are not very blockspace efficient yet.

```suggestion            
```

nopara73 commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletView.xaml:
```diff
@@ -29,6 +29,12 @@
+          <CheckBox IsChecked="{Binding ShowAdvancedOptions,Mode=TwoWay}"><TextBlock Text="Show advanced options"/></CheckBox>
+          <StackPanel Spacing="8" IsVisible="{Binding ShowAdvancedOptions}">
+            <controls:ExtendedTextBox Text="{Binding AccountKeyPath}" Watermark="Account Key Path" UseFloatingWatermark="True"/>
+            <controls:ExtendedTextBox Text="{Binding MinGapLimit}" Watermark="Number of keys to derive" UseFloatingWatermark="True"/>
```
```suggestion
            <controls:ExtendedTextBox Text="{Binding MinGapLimit}" Watermark="Minimum Gap Limit" UseFloatingWatermark="True"/>
```

We occasionally derive more than the minimum gap limit, thus minimum and gap limit, because the "keys to derive" would suggest for me that this many keys would only be derived the first time around.

nopara73 commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletView.xaml:
```diff
@@ -29,6 +29,12 @@
               </DataTemplate>
             </ItemsControl.ItemTemplate>
           </ItemsControl>
+          <CheckBox IsChecked="{Binding ShowAdvancedOptions,Mode=TwoWay}"><TextBlock Text="Show advanced options"/></CheckBox>
```
Formatting. Sorry, from next version we'll want to fix up the code smells and formatting inconsistencies.
```suggestion
          <CheckBox IsChecked="{Binding ShowAdvancedOptions, Mode=TwoWay}">
              <TextBlock Text="Show advanced options"/>
          </CheckBox>
```

nopara73 replied:
There should also be a spacing/padding or something here: 8 I think. Did you try if it looks a bit better?

nopara73 commented:
One last thing for the UX, the `MinGapLimit` should have somehow an enforced minimum: `AbsoluteMinGapLimit` and some enforced maximum, like `1000000`(1 million).

@guggero Would you like to work on our recommendations or you would prefer if we take over from here?

nopara73 commented:
> I hope that's in the general interest of the project since people might want to import their existing wallet rather than create a new one and send funds to it.

Ok, I see we are on the same page here:)

nopara73 commented:
Oh, one more thing. We can only monitor bech32 addresses. So the user may get confused if he writes `m/44'/...` and doesn't see his balance.

nopara73 commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletView.xaml:
```diff
@@ -29,6 +29,12 @@
           </ItemsControl>
+          <CheckBox IsChecked="{Binding ShowAdvancedOptions,Mode=TwoWay}"><TextBlock Text="Show advanced options"/></CheckBox>
+          <StackPanel Spacing="8" IsVisible="{Binding ShowAdvancedOptions}">
+            <controls:ExtendedTextBox Text="{Binding AccountKeyPath}" Watermark="Account Key Path" UseFloatingWatermark="True"/>
```
I think we need a note here. We use the "Note that, " language syntax in the Send tab so it'd be consistent like this.

```suggestion
            <controls:ExtendedTextBox Text="{Binding AccountKeyPath}" Watermark="Account Key Path" UseFloatingWatermark="True"/>
            <TextBlock Text="Note that, Wasabi can only monitor bech32 addresses."/>
```

guggero commented:
@nopara73 thanks a lot for your feedback!
I'm happy this is a feature you agree with.
I will gladly implement your suggestions. It will be a busy week for me so it might take me a few days.
There seem to be build errors too, so I'll fix them as well.

guggero commented:
I've rebased against master and addressed all your comments/suggestions. The test that I broke should now run fine too.
The screenshot shows how it currently looks.

lontivero commented:
I will review this tomorrow.

lontivero commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletView.xaml:
```diff
@@ -29,6 +29,18 @@
+              <TextBlock Text="Note that, Wasabi can only monitor native SegWit (bech32) addresses."/>
+            </StackPanel>
+            <StackPanel IsVisible="{Binding ShowAdvancedOptions}">
+              <controls:ExtendedTextBox Text="{Binding MinGapLimit}" Watermark="Minimum Gap Limit" UseFloatingWatermark="True"/>
```
It could worth to use a `NumericUpDown` control for the MinGapLimit because it makes easier to limit min and max and also make sure that a numerical value is entered.

guggero replied:
I tried that. But then we don't have that nice floating watermark. And I noticed that if the property `MinGapLimit` is of type `int` then the ExtendedTextBox automatically parses the entered text as a number and also shows default error messages. So we only need to check upper and lower bound as is done in `RecoverWalletViewModel`.
So I didn't change it at the moment, hope that's ok.

lontivero replied:
okay.

lontivero commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletViewModel.cs:
```diff
@@ -46,17 +49,34 @@ public RecoverWalletViewModel(WalletManagerViewModel owner) : base("Recover Wall
 					}
+					catch (FormatException)
+					{
+						ValidationMessage = "The account key path is not a valid derivation path.";
```
The error message is misleading. Here we can get a `FormatException` for several reasons, not just for invalid derivation path. For example, if the number of mnemonics is wrong you will get this exception too and the user will review the derivation path.

My recommendation is to create a `bool TryParseKeyPath(string, KeyPath)` and use that for the moment. Next, wrap the `Recover` around a try-catch(Exception) block.

lontivero commented on WalletWasabi.Gui/Tabs/WalletManager/RecoverWalletViewModel.cs:
```diff
@@ -157,6 +195,9 @@ public override void OnCategorySelected()
 			WalletName = Utils.GetNextWalletName();
 			ValidationMessage = null;
+			ShowAdvancedOptions = false;
+			AccountKeyPath = $"m/{KeyManager.DefaultAccountKeyPath.ToString()}";
```
`.ToString()` is redundant. Remove it, please.

lontivero commented on WalletWasabi/KeyManagement/KeyManager.cs:
```diff
@@ -69,10 +75,11 @@ public class KeyManager
 
-		private const int AbsoluteMinGapLimit = 21;
+		public const int AbsoluteMinGapLimit = 21;
+		public const int AbsoluteMaxGapLimit = 1_000_000;
```
Uhmm... if someone sets the `minGapLimit` to a 1e6 it could take a long time to derivare that number of keys. Have you measured how much it takes?

guggero replied:
No, I didn't try it myself. @nopara73 suggested to add the upper bound of 1 million. It seemed large to me too but who am I to disagreee ;-)

lontivero commented on WalletWasabi/KeyManagement/KeyManager.cs:
```diff
@@ -141,8 +150,9 @@ public static KeyManager CreateNew(out Mnemonic mnemonic, string password, strin
 			HDFingerprint masterFingerprint = extKey.Neuter().PubKey.GetHDFingerPrint();
-			ExtPubKey extPubKey = extKey.Derive(AccountKeyPath).Neuter();
-			return new KeyManager(encryptedSecret, extKey.ChainCode, masterFingerprint, extPubKey, false, AbsoluteMinGapLimit, new BlockchainState(), filePath);
+			KeyPath keyPath = DefaultAccountKeyPath;
```
I think the variable is redundant here.

nopara73 commented:
@lontivero @guggero Scream when you guys think it's ready and I'll do some final fixup before merging.

guggero commented:
Thanks @lontivero for the review!
I addresses most of your comments. Sorry I only found time to do so today.
If you agree with my comments @nopara73, I think the PR is ready to merge. Otherwise I'm happy to change the max value for `MinGapLimit` and/or use a `NumericUpDown`.

lontivero commented:
I think what needs to be addressed is https://github.com/zkSNACKs/WalletWasabi/pull/1376#discussion_r280067449

guggero commented:
@lontivero I have addressed that, there is a `TryParseKeyPath` function now.
I've also rebased to master.
@nopara73 I think this might be ready to be merged.

nopara73 commented:
@guggero Awesome! I am on it.

nopara73 commented:
I removed the maxgaplimit from the KeyManager, and made it a UI concept, because I don't want to pay attention to the max gap limit inside the KeyManager: https://github.com/zkSNACKs/WalletWasabi/pull/1376/commits/7c46b4c08a6e8e715895e1d1b905e05f87e4a907

It's also not max gap limit, but max min gap limit:)

nopara73 commented:
I added a custom keypath converter. NBitcoin's doesn't always work in the way we want.

guggero commented:
Cool, thanks a lot!

nopara73 commented:
What's with the trickering of MinGapLimit +- 1? I don't understand why that's needed.

nopara73 commented:
> What's with the trickering of MinGapLimit +- 1? I don't understand why that's needed.

After we discussed this, it can be merged.

guggero commented:
Because in the KeyManager it's always `.Count() < MinGapLimit` and the default was 21 before. And at first I named it "number of keys to generate" and 21 wouldn't have been correct.
But with the wording you suggested it's not needed any more.

guggero commented:
I've removed the now unnecessary one-off corrections of `MinGapLimit`.

nopara73 commented:
Awesome! Merging. Although we have a feature freeze (https://github.com/zkSNACKs/WalletWasabi/issues/1412), I make an exception for this PR, this looks really safe.

guggero commented:
Thanks for merging so quickly! And thanks a lot for making an exception in the feature freeze :-)

MaxHillebrand commented:
I found a little bug, not sure if I should open a new issue...

When typing in the mnemonic, and you have a typo, you can not move the cursor back to the misspelled word, neither by mouse clicking, nor by arrow key press.

When you click and select a word, and start typing, then the cursor jumps back to the end of the mnemonic and the letters are then there.

MaxHillebrand commented:
Since this specific PR is for only the advanced option, I guess it's best to open a [new issue](https://github.com/zkSNACKs/WalletWasabi/issues/1429).
There is a [closed issue](https://github.com/zkSNACKs/WalletWasabi/issues/775) with the same bug.

Sorry for spamming this PR...
